sys/include/net/dev_eth.h

```diff
@@ -0,0 +1,198 @@
+     * @param len nr of bytes to send
+     *
+     * @return nr of bytes sent, or <=0 on error
+     */
+    int (*send)(dev_eth_t *dev, char* buf, int len);
+
+    /**
+     * @brief Get a received ethernet frame
+     *
+     * Supposed to be called from dev_eth_rx_handler().
+     *
+     * Make sure buf can hold the maximum expected ethernet frame size.
+     *
+     * @param buf buffer to write to
+     * @param len maximum nr. of bytes to read
+     *
+     * @return nr of bytes read or <=0 on error
+     */
+    int (*recv)(dev_eth_t *dev, char* buf, int len);
+
+    /**
+     * @brief get the device's MAC address
+     *
+     * @param buf location to write to. Make sure this is can take least 6 bytes
+     */
+    void (*get_mac_addr)(dev_eth_t *dev, uint8_t *buf);
+
+    /**
+     * @brief get the device's promiscous status
+     *
+     * Default is not promiscous.
+     * Promiscous means, receive all ethernet frames.
+     * Not promiscous means only receive broadcast, multicast or frames with
+     * the device's MAC address as dst address.
+     *
+     * @return 1 for promiscous, 0 for not promiscous
+     */
+    int (*get_promiscous)(dev_eth_t *dev);
+    /**
+     * @brief set the devices promiscous mode
+     *
+     * @param value 1 for promiscous, 0 for not promiscous
+     * @return the new value (device might not support wanted mode)
+     */
+    int (*set_promiscous)(dev_eth_t *dev, int value);
+    /**
+     * @brief the driver's initialization function
+     *
+     * @return <=0 on error, >0 on success
+     */
+    int (*init)(dev_eth_t *dev);
+    /**
+     * @brief a driver's user-space ISR handler
+     *
+     * This function will be called from a network stack's loop when being notified
+     * by dev_eth_isr.
+     *
+     * It is supposed to call dev_eth_rx_handler for each available received packed,
+     * and dev_eth_linkstate_handler whenever a link state change event occurs.
+     *
+     * See receive packet flow description for details.
+     */
+    void (*isr)(dev_eth_t *dev);
+} eth_driver_t;
+
+/**
+ * @brief Initialize a device given by dev (convenience function)
+ *
+ * The device given as parameter *must* be previously setup by the
+ * drivers *_setup() function.
+ *
+ */
+static inline int dev_eth_init(dev_eth_t *dev) {
+    return dev->driver->init(dev);
+}
+
+/**
+ * @brief global dev_eth interrupt handling function.
+ *
+ * This function should be called from your device's ISR from whithin ISR context.
+ * It is supposed to wake up a waiting user-space event loop.
+ */
+extern void dev_eth_isr(dev_eth_t *dev);
+
+/**
+ * @brief dev_eth event callback for packets that were received.
+ *
+ * This function should be called from whithin your driver's isr()
+ * (as defined in eth_driver_t), once for every packet that was received.
+ *
+ * It needs to call dev->driver->recv() in order to get received packet
+ * from the driver.
+ */
+extern void dev_eth_rx_handler(dev_eth_t *dev);
```

Comment:
The 'dev_eth_isr()' function lets the driver loop know that the ISR was triggered and there's work to do. It is called from interrupt context, and lets a thread know that driver->isr has to be run.

This driver->isr then calls dev_eth_rx_handler or the linkstate handler.